src/src/com/microsoft/aad/adal/HttpWebRequest.java

```diff
@@ -107,107 +92,116 @@ private void setupConnection() {
             throw new IllegalArgumentException("requestURL");
         }
         HttpURLConnection.setFollowRedirects(true);
-        mConnection = openConnection();
+        final HttpURLConnection connection = (HttpURLConnection)mUrl.openConnection();
+        connection.setConnectTimeout(CONNECT_TIME_OUT);
         // To prevent EOF exception.
         if (Build.VERSION.SDK_INT > 13) {
-            mConnection.setRequestProperty("Connection", "close");
+            connection.setRequestProperty("Connection", "close");
+        }
+
+        // Apply the request headers
+        final Iterator<String> headerKeys = mRequestHeaders.keySet().iterator();
+
+        while (headerKeys.hasNext()) {
+            String header = headerKeys.next();
+            Logger.v(TAG, "Setting header: " + header);
+            connection.setRequestProperty(header, mRequestHeaders.get(header));
         }
+
+        // Avoid reuse of existing sockets to avoid random EOF errors
+        System.setProperty("http.keepAlive", "false");
```

Comment:
Actually this will disabled connection reuse for every client that uses a library. For example for us it would have a significant impact on thumbnail downloading code...